### ISSUE TITLE
Add experimental support for post processors

### ISSUE THREAD

**Issue by YuanboXue-Amber** [stylelint/stylelint#7567]
### What is the problem you're trying to solve?

I'm creating a stylelint custom syntax for a css-in-js engine. While lint result from `stylelint.lint` successfully identifies errors/warnings in css, the result is based on their location in css AST. And I would like to post process these locations into original locations in the js code (and therefore have correct squiggly lines from vscode extension).

Btw in ESlint, there's [custom processors](https://eslint.org/docs/latest/extend/custom-processors) `preprocess` which handles similar situation as Stylelint `customSyntax`. And ESlint `postprocess` allows for adjusting the linting result. I have not found a similar feature in Stylelint that would allow for a `postprocess` - like manipulation of the linting result object.

### What solution would you like to see?

A config option in style lint. Something like a callback `(lintResult: Promise<stylelint.LinterResult>) => Promise<stylelint.LinterResult>`. And `stylelint.lint` will call this callback at the very end, allowing post process of the lint result.

If this request makes sense, I would be happy to contribute :) - edit: here's a draft implementation #7568

**Comment by ybiquitous** [comment]
@YuanboXue-Amber Thanks for the request. I still need to understand the necessity of such post-processors.

There are already some CSS-in-JS custom syntaxes like [`postcss-styled-syntax`](https://www.npmjs.com/package/postcss-styled-syntax) (see also the [awesome list](https://stylelint.io/awesome-stylelint/#custom-syntaxes)), but do these syntaxes also need post-processors?

**Comment by YuanboXue-Amber** [comment]
@ybiquitous for
>the necessity of such post-processors

Thank you for your interest in understanding the need for a postprocessor in our use case. Allow me to elaborate on the specifics:

we need a post processor because our css-in-js engine does module evaluation. For example we can have below js:
```js
// in sharedStyles.js
export const sharedStyles = { backgroundColor: "#000" };

// sharedStyles.js is imported and used in styles.js
import { makeStyles } from "@griffel/react";
import { sharedStyles } from "./sharedStyles.js";
export const useStyles = makeStyles({
  container: {
    ...sharedStyles,
    color: "#fffff",
  },
});
```
And in our custom post-css syntax, we evaluate the imports and transform `style.js` to AST of the below css:
```css
.class1 { background-color: #000 } 
.class2 { color: #fff}
```
#### With post processor
PostCSS AST allows us to store extra raw data, for instance we could annotate that `.class1` originated from the `sharedStyles` node. And postprocessor allows us to accurately remap error/warning in .class1 rule to the corresponding location of `sharedStyles` node in JS.

#### Without post processor 
Without a postprocessor, our only option is to manipulate the start/end locations within the PostCSS AST. This approach is not only complex due to the module evaluation but also limited in precision. For example, an error/warning in the CSS rule `#000` would be 4-character length. Consequently, the best we could achieve is aligning the linting annotations with the first 4-characters of `...sharedStyles` in the JS source, which is not ideal.

**Comment by YuanboXue-Amber** [comment]
`customSyntax` is great, but it is more of a "pre-processor" and we would like the ability to "post-process" the lint result.

**Comment by ling1726** [comment]
@ybiquitous - this feature is necessary for CSS syntaxes that is not 'standard' css or postcss. In our case we need it for css-in-js. So in fact the file being parsed is javascript.

The way stylelint marks error row/col values makes an inherent assumption that it is linting CSS or SASS variants. When we extend to javascript it's no longer possible to make the same assumptions because as @YuanboXue-Amber  pointed out, there's no guarantee you have a defined property key value pair.

This is definitely and edge case so there shouldn't be any core stylelint changes to handle javscript. However a bailout where we users can manually post process the result to would be really helpful

**Comment by ybiquitous** [comment]
@YuanboXue-Amber @ling1726 Thank you for your elaboration. I understand your motivation so well. 👍🏼 

I agree with adding the post-processing feature since the addition should keep compatibility. However, I think we should consider the API design before implementing the feature because we previously provided a similar feature for pre-processing. See https://github.com/stylelint/stylelint/blob/14.16.1/docs/developer-guide/processors.md.

Or, we may be able to ship the feature as experimental to collect user feedback and avoid unnecessary breaking changes.

Any thoughts?

**Comment by YuanboXue-Amber** [comment]
@ybiquitous I'm open to suggestion on different API. I created PR #7568 which proposes `postProcessor: (results: stylelint.LintResult[]) => stylelint.LintResult[]`. This proposal stems from our need to access PostCss AST and update the locations in lint results.

Also I'm open to shipping experimental feature. My team is planning to have stylelint on a large codebase in microsoft. If it is decided that experiment feature is preferable, I would appreciate being tagged or notified in case the feature gets removed in the future.

**Comment by YuanboXue-Amber** [comment]
I like it! It makes more sense to have an object instead of just a function. I will update my PR to align with this proposal

**Comment by ling1726** [comment]
@ybiquitous  my 2cents

> Support a single processor or multiple processors (processor vs processors)
> Processor function's type for arguments and a return value

IMO multi processor is totally valid but that brings up the point of what should the signature be? should the processor API return a new set of results of simply mutate the existing one passed in arguments?


There's another side I just thought of here - what about [ignoring code/rules](https://stylelint.io/user-guide/ignore-code/)? How does that currently work in stylelint? is it based on parsing comments in the postcss syntax?

**Comment by Mouvedia** [comment]
> However, I think we should consider the API design _before_ implementing the feature…

👍

**Comment by ybiquitous** [comment]
@YuanboXue-Amber 

> It makes more sense to have an object instead of just a function.

The idea of a processor function comes from Stylelint's [removed processor API](https://github.com/stylelint/stylelint/blob/14.16.1/docs/user-guide/configure.md#processors). A processor function can change its behavior by the `options` argument passed the function, e.g.

```js
export default function myProcessor(options = {}) {
  return {
    name: 'my-processor',

    postprocess(results) {
      if (options.a) {
        // ...
      } else if (options.b) {
        // ...
      }
    }
  };
};
```

I feel this way could provide more flexibility, but it may be too much as the first implementation.

**Comment by ybiquitous** [comment]
@ling1726 

> IMO multi processor is totally valid but that brings up the point of what should the signature be? should the processor API return a new set of results of simply mutate the existing one passed in arguments?

Good point. Whether mutable or immutable is a topic of discussion.

The current implementation operates lint results mutably, so I feel mutable is more natural and consumes more performant because of no copied objects. 

https://github.com/stylelint/stylelint/blob/2d74e26c93b69793d7375d25956a218818ee2601/lib/prepareReturnValue.mjs#L19-L28

However, mutable operations may have downsides, while immutability may benefit. Please let us know if you have more information.

**Comment by ybiquitous** [comment]
@ling1726 

> There's another side I just thought of here - what about [ignoring code/rules](https://stylelint.io/user-guide/ignore-code/)? How does that currently work in stylelint? is it based on parsing comments in the postcss syntax? We're writing processors to make sure the errors are relevant to code, but what is the work needed to make sure disabling comments aligns with the result 🤔🤔

The configuration comments like `/* stylelint-disable */` removes lint problems from results. Also, such comments are processed as a PostCSS's [Comment](https://postcss.org/api/#comment) object. The following code may help your understanding:

https://github.com/stylelint/stylelint/blob/2d74e26c93b69793d7375d25956a218818ee2601/lib/assignDisabledRanges.mjs#L288-L291

**Comment by ling1726** [comment]
> The configuration comments like `/* stylelint-disable */` removes lint problems from results. Also, such comments are processed as a PostCSS's [Comment](https://postcss.org/api/#comment) object. The following code may help your understanding:
> 
> https://github.com/stylelint/stylelint/blob/2d74e26c93b69793d7375d25956a218818ee2601/lib/assignDisabledRanges.mjs#L288-L291

I would be interested to explore that once @YuanboXue-Amber's changes are public (whether is full or unstable), to figure out what the edge cases would be once we are able to map these errors to different syntaxes more accurately.

**Comment by YuanboXue-Amber** [comment]
I updated #7568 to according to [the proposal](https://github.com/stylelint/stylelint/issues/7567#issuecomment-2020365221) by @ybiquitous, with one change:
from 
```js
    /** @type {(results: LintResult[]) => void} */
    postprocess(results) {
      // Mutate results...
    }
```
to
```js
    /** @type {(results: LintResult) => void} */
    postprocess(result) {
      // Mutate results...
    }
```

I made this change because the `results` array is just `[result]`.

Appreciate reviews/feedbacks 🙏

**Comment by ybiquitous** [comment]
@YuanboXue-Amber Thanks for the pull request.

I suggest implementing the feature as an experiment by documenting it because I still wonder whether this proposed API may have any downsides. 🤔 

For example, adding a note as below to the processors' section:

> [!WARNING]
> This is an experimental feature. The behavior is subject to change.

**Comment by ybiquitous** [comment]
Since it seems that there are no concerns, I just transitioned this issue to `status: wip`. Even if a problem is found after release, we could change or remove the experimental feature.

**Comment by ybiquitous** [comment]
@ling1726 I'll release it today. Please be patient.

**Comment by Mouvedia** [comment]
@ybiquitous the next time that we release an experimental feature, should we have [ExperimentalWarning](https://nodejs.org/api/process.html#nodejs-warning-names) in place beforehand?

**Comment by ybiquitous** [comment]
@Mouvedia Good catch, I didn't know the `ExperimentalWarning` type.

Could you open an issue to add the type? We probably could extend the existing [`emitDeprecationWarning()`](https://github.com/stylelint/stylelint/blob/a0c6dc43d32f3c93e7bcef090c06aefd59373e08/lib/utils/emitDeprecationWarning.mjs) utility.
